### PR TITLE
#10218 - Fix for session regenerate id issue after user logged out

### PR DESCRIPTION
**Pull request by andrewnester**
This is fix for #10218 
It is based on solution @klimov-paul offered in conversation for this ticket

### CODE REVIEW

**Comment by SilverFire** [comment]
This fix does not work on my env `PHP 5.6.11-1ubuntu3.1`

```php
    public function actionLogout()
    {
        $id0 = Yii::$app->session->getId(); // g0c21rtignr91r9353asddg370
        Yii::$app->user->logout();
        $id1 = Yii::$app->session->getId(); // '' (empty string)
        Yii::$app->session->setFlash('success', 'You have been successfully logged out!');
        $id2 = Yii::$app->session->getId(); // '' (empty string)

        return $this->goHome();
    }
```

And then session data will be never saved correctly. I don't have a clue, at least for now.

**Comment by SilverFire** [comment]
@klimov-paul how did it work for you?

**Comment by andrewnester** [comment]
Hi @SilverFire 
You were absolutely right, it didn't work as expected.
I've just implemented new solution, tested it on basic application - looks like it works fine.
Now after `Session::destroy` called, new session with new session id is started

**Comment by SilverFire** [comment]
Travis failed on PHP7

**Comment by andrewnester** [comment]
Fixed, sorry. All checks are passed now. Looks like session_start behavior was changed in PHP7.

File: framework/web/Session.php [hunk 1]
```diff
@@ -183,6 +183,8 @@ public function destroy()
         if ($this->getIsActive()) {
             @session_unset();
             @session_destroy();
+            @session_start();
```
Comment (klimov-paul) [inline framework/web/Session.php]: This is obviously incorrect: `session_start()` will start new session, thus there is no way to close session anymore.

**Comment by klimov-paul** [comment]
> It is based on solution @klimov-paul offered in conversation for this ticket

Well, I don't see my proposal here. Back in #10218 I suggested setting session ID to `null` manually on destroy: 

```php
@session_id(null);
```

Using of `session_regenerate_id()` or especially `session_start()` inside `destroy()` is not acceptable.
If my original solution is not correct by some reason, we should leave the code as it is and update the documentation, warning the developers they may need to invoke `Session::regenerateId()` manually.

**Comment by SilverFire** [comment]
> Well, I don't see my proposal here

It was implemented initially, but then @andrewnester did a force push with update, I guess

**Comment by SilverFire** [comment]
> Using of session_regenerate_id() or especially session_start() inside destroy() is not acceptable.

Agree on this.

**Comment by andrewnester** [comment]
@klimov-paul yes, @SilverFire is right, previously it was using `session_id(null)`, sorry for forced update. it was changed because `session_id(null)` doesn't solve the problem with flash messages, at least for PHP 5.6 and PHP 5.5 (tested on couple of versions).

I will try to investigate further how it could be solved without starting new session

**Comment by andrewnester** [comment]
I just found and implemented solution based on this comment http://php.net/manual/en/function.session-destroy.php#114709
What do you think? PR has been updated too

**Comment by klimov-paul** [comment]
I am afraid this relates to the different matter and unlikely can solve the problem.

**Comment by andrewnester** [comment]
@klimov-paul I've tested it on basic app with PHP 5.6.16 and PHP 5.5.30 - looks like works fine and flash messages successfully displayed after logging out

**Comment by klimov-paul** [comment]
Well, update PR - lets have a look on it.

**Comment by SilverFire** [comment]
It's already updated

**Comment by klimov-paul** [comment]
Looks fine.

**Comment by SilverFire** [comment]
```php
    public function actionLogout()
    {
        $id0 = Yii::$app->session->getId(); // fqv0q23h6870fdc9esmvdcpri2
        Yii::$app->user->logout();
        $id1 = Yii::$app->session->getId(); // v3apcm77672sbrof45onm0m9q4
        Yii::$app->session->setFlash('success', 'You have been successfully logged out!');

        return $this->goHome();
    }
```

Session id changes, but the bug seems to be fixed

**Comment by andrewnester** [comment]
Yes, it is expected, because call to `User::logout` that calls `User::switchIdentity` actually creates new session. The most important thing for fix is that after `Session::destroy`, session id must remain the same.

**Comment by SilverFire** [comment]
Merged, thank you!

**Pull request merged**